### PR TITLE
crimson/osd: don't ignore start_pg_operation returned future

### PR DESCRIPTION
**Pull request by Matan-B**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/osd.cc [hunk 1]
```diff
@@ -1016,17 +1016,17 @@ seastar::future<> OSD::committed_osd_maps(version_t first,
 seastar::future<> OSD::handle_osd_op(crimson::net::ConnectionRef conn,
                                      Ref<MOSDOp> m)
 {
-  (void) pg_shard_manager.start_pg_operation<ClientRequest>(
+  return pg_shard_manager.start_pg_operation<ClientRequest>(
```
Comment (cyx1231st) [inline src/crimson/osd/osd.cc:1019]: LGTM as these future won't be chained to `ms_dispatch()`:
https://github.com/ceph/ceph/blob/43bf04e6e9cb82fa7cfe5644786d91e6265bcc8a/src/crimson/osd/osd.cc#L696-L707

**Comment by liu-chunmei** [comment]
@Matan-B what problem did the ignore future cause?

**Comment by Matan-B** [comment]
> @Matan-B what problem did the ignore future cause?

No issues fixed here (yet). It may be useful for a future PR.
Added cleanup label as this is not a bug.

File: src/crimson/osd/osd.cc (hunk 1 above)
Comment (Matan-B) [reply src/crimson/osd/osd.cc:1019]: Do you refer to the ignored future when calling dispatch?
```
inline void dispatch_in_background(const char* what, T& who, Func&& func) {
  (void) dispatch(what, who, func);
}
```

Comment (cyx1231st) [reply src/crimson/osd/osd.cc:1019]: Yes, see https://github.com/ceph/ceph/blob/main/src/crimson/net/Dispatcher.h#L27-L30. `ms_dispatch()` should not return an unavailable future unless there is a need about throttling. This is why it always returns `seastar::now()` if dispatched.

> Do you refer to the ignored future when calling dispatch?

The future is correct to be "ignored" here and be gated instead, which is the exact meaning of "dispatch in background".

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
jenkins test api

**Comment by Matan-B** [comment]
jenkins test api

**Comment by athanatos** [comment]
It sounds like we specifically want to drop the future unless we are throttling.  Seems like this should be closed?

**Comment by cyx1231st** [comment]
> It sounds like we specifically want to drop the future unless we are throttling.

I was double checking the code is still correct that won't throttle the messenger through ms_dispatch().

>  Seems like this should be closed?

Seems not to me.

This PR fixes a different issue to not ignore any future from `start_pg_operation()` and use OSD gate to track them in the background.

**Comment by athanatos** [comment]
Oh, got it.

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
```
Merge remote-tracking branch 'wip-crimson-ertr-plain-rets' into wip-matanb-crimson-testing-28.6
Merge remote-tracking branch wip-bug-59429' into wip-matanb-crimson-testing-28.6
Merge remote-tracking branch 'wip-matanb-crimson-watch-notify-logs' into wip-matanb-crimson-testing-28.6
Merge remote-tracking branch 'wip-matanb-crimson-no-ignore' into wip-matanb-crimson-testing-28.6
latest main commit 0082f63e8c
```
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-28.6

Build is ready, awaiting infra failures to resolve.

**Comment by Matan-B** [comment]
Merging based on:
https://pulpito.ceph.com/matan-2023-07-02_07:05:40-crimson-rados-wip-matanb-crimson-testing-28.6-distro-crimson-smithi/

**Pull request merged**